File: Python/PASTRY_sim.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse, time
import numpy as np
import scipy.stats as stats
from numpy import savetxt
from numpy import random as np_random
import pandas as pd


def pastry_sim(RRs, maf, nrep, k, ncase, ncont, ncontshared, t):

    nsnp=k
    betas=np.zeros((k,nsnp))

    unit=np.ones(k)
    unit_t=np.transpose(unit)
    preval=0.000

    RRs=np.repeat(RRs, k)
    nstudy=np.repeat(ncont+ncase,k)
    ncont=np.repeat(ncont,k)
    ncase=np.repeat(ncase,k)
    ncontspe=ncont-np.repeat(ncontshared,k)
    contnsplit=ncontshared/k
    
    print(" * Check the parameter : ")
    print("  - Studies # for meta-analysis : ", k)
    print("  - Risk Ratio : ",RRs)
    print("  - MAF : ",maf)
    print("  - Simulation # : ",nrep)
    print("  - Case # : ",ncase)
    print("  - Non-shared cont # : ",ncontspe)
    print("  - Shared cont # : ",ncontshared)
    print("  - Threshold : ",t,"\n")

    ## Case/control frequencies
    casemaf=(RRs*maf) / ((RRs-1)*maf + 1)
    contmaf=(maf-preval*casemaf)/(1-preval)

    ## Generating Correlation Matrix
    ls_R=np.diagflat(np.repeat(1,k)).astype(float)
    
    for i in range(k):
        for j in range(k):
            if i!=j:
                num=ncontshared*np.sqrt((ncase[i]*ncase[j])/(ncont[i]*ncont[j]))
                den=np.sqrt(nstudy[i]*nstudy[i])
                ls_R[i,j]=num/den
	
    inv_ls_R=np.linalg.inv(ls_R)

    ## FE subroutine
    def FE(beta, stders):
        inv_var=stders**(-2)
        inv_var_beta=(np.dot(inv_var, beta)/sum(inv_var))
        inv_var_beta_var=1/sum(inv_var)
        z=inv_var_beta/np.sqrt(inv_var_beta_var)
        p_FE=stats.norm.pdf(abs(z))*2
        return(pd.DataFrame(data={'z':[z],'p':[p_FE],'beta':[inv_var_beta],'se':[np.sqrt(inv_var_beta_var)]}))

    pval_LS_FE=np.zeros(nrep)
    pval_PASTRY_FE=np.zeros(nrep)
    pval_split_FE=np.zeros(nrep)

    # k = 2 two cohort study
    i = 1; j = 2
	
    ## Now generate studies
    record=2
    beta_nonde=[]
    stderr_nonde=[]
    beta_split=[]
    stderr_split=[]

    for eachrep in range(nrep):
        betas_nonde=np.zeros(k)
        stder_nonde=np.zeros(k)
        betas_split=np.zeros(k)
        stder_split=np.zeros(k)
        casemac=np.zeros(k)
        contmac=np.zeros(k)
        contsplitmac=np.zeros(k)

        for i in range(k):
            #case-specific
            casemac[i]=np.random.binomial(n=ncase[i]*2, p=casemaf[i], size=1)
            #cont-specific
            contmac[i]=np.random.binomial(n=ncontspe[i]*2, p=contmaf[i], size=1)
            #cont-shared
            contsplitmac[i]=np.random.binomial(n=2*contnsplit, p=contmaf[i], size=1)

        contsharedmac=np.sum(contsplitmac)

        for i in range(k):
            ## non-splitting
            n11=casemac[i]
            n10=2*ncase[i]-casemac[i]
            n01=contmac[i]+contsharedmac
            n00=2*ncontspe[i]+2*ncontshared-contmac[i]-contsharedmac
            betas_nonde[i]=np.log(n11*n00/n01/n10)
            stder_nonde[i]=np.sqrt(1/n11 + 1/n00 + 1/n01 + 1/n10)
            beta_nonde.append(betas_nonde[i])
            stderr_nonde.append(stder_nonde[i])

            ## splitting
            n11=casemac[i]
            n10=2*ncase[i]-casemac[i]
            n01=contmac[i]+contsplitmac[i]
            n00=2*ncontspe[i]+2*contnsplit-contmac[i]-contsplitmac[i]
            betas_split[i]=np.log(n11*n00/n01/n10)
            stder_split[i]=np.sqrt(1/n11 + 1/n00 + 1/n01 + 1/n10)
            beta_split.append(betas_split[i])
            stderr_split.append(stder_split[i])

        ## Lin
        i = 1; j = 2

        p0cont=1-contmaf[i]
        p1cont=contmaf[i]
        p0case=1-casemaf[i]
        p1case=casemaf[i]

        invstdmat=np.linalg.pinv(np.diagflat(stder_nonde))
        ls_invsigma=np.dot(np.dot(invstdmat,inv_ls_R),invstdmat)
        ls_invvar=np.dot(np.dot(unit,ls_invsigma),unit_t)
        ls_beta_num=np.dot(np.dot(unit,ls_invsigma),betas_nonde)
        ls_beta=ls_beta_num/ls_invvar
        ls_var=1/ls_invvar
        ls_stat=ls_beta/np.sqrt(ls_var)
        pval_LS_FE[eachrep]=stats.norm.pdf(abs(ls_stat))*2

        if eachrep == 1:
           print(" * PASTRY statistics are generating... for each SNP. \n")
           start=time.time()

        ntotali=ncont[i]+ncase[i]
        ntotalj=ncont[j]+ncase[j]
        
        exp_alpha_i = np.exp(np.log(ncase[i]/(ncont[i]))-(ls_beta*(((ncase[i]*casemaf[i])+(ncont[i]*contmaf[i]))/ntotali)))
        exp_alpha_j = np.exp(np.log(ncase[j]/(ncont[j]))-(ls_beta*(((ncase[j]*casemaf[j])+(ncont[j]*contmaf[j]))/ntotalj)))

        x0_i = exp_alpha_i
        x1_i = exp_alpha_i*np.exp(ls_beta)

        x0_j = exp_alpha_j
        x1_j = exp_alpha_j*np.exp(ls_beta)

        ## Information Matrix
        x0=np.array([1,0,0,0]).reshape(2,2)
        x1=np.array([1,1,1,1]).reshape(2,2)

        up_covmat=np.array([0,0,0,0]).reshape(2,2)

        info_i=2*ncase[i]*(p0case*(x0_i/(1+x0_i)**2)*x0 + p1case*(x1_i/(1+x1_i)**2)*x1) + 2*(ncont[i])*(p0cont*(x0_i/(1+x0_i)**2)*x0 + p1cont*(x1_i/(1+x1_i)**2)*x1)
        info_j=2*ncase[j]*(p0case*(x0_j/(1+x0_j)**2)*x0 + p1case*(x1_j/(1+x1_j)**2)*x1) + 2*(ncont[j])*(p0cont*(x0_j/(1+x0_j)**2)*x0 + p1cont*(x1_j/(1+x1_j)**2)*x1)

        ### Covarinace Matrix for PASTRY
        cov00=p0cont*(x0_i*x0_j/((1+x0_i)*(1+x0_j)))*x0
        cov11=p1cont*(x1_i*x1_j/((1+x1_i)*(1+x1_j)))*x1

        covmatrix=2*ncontshared*(cov00+cov11)

        inv_i = np.linalg.pinv(info_i)
        inv_j = np.linalg.pinv(info_j)

        covfinal=np.dot(np.dot(inv_i,covmatrix),inv_j)

        ### Correlation matrix for PASTRY
        cor=covfinal[1,1]/np.sqrt(np.dot(inv_i[1,1],inv_j[1,1]))
        pastry_R=np.tile(cor,(k,k))
        np.fill_diagonal(pastry_R, 1)

        ## new decoupling
        pastry_invsigma=np.dot(np.dot(invstdmat,np.linalg.inv(pastry_R)),invstdmat)
        pastry_invvar=np.dot(np.dot(unit,pastry_invsigma),unit_t)
        pastry_var=1/pastry_invvar
        beta_num=np.dot(np.dot(unit,pastry_invsigma),betas_nonde)	
        pastry_beta=beta_num/pastry_invvar
        pastry_stat=pastry_beta/np.sqrt(pastry_var)

        ## pvalue calculation by FE
        pval_split_FE[eachrep]=FE(betas_split, stder_split)['p']
        pval_PASTRY_FE[eachrep]=stats.norm.pdf(abs(pastry_stat))*2

    end=time.time()
    print(" * # Simulation : ",nrep," \n * Simulation Results : \n  LS : ",sum(pval_LS_FE<t)/nrep,", PASTRY",sum(pval_PASTRY_FE<t)/nrep,", split_FE",sum(pval_split_FE<t)/nrep,"\n")
    print("Done.\n")
    print(" * PASTRY Simulation running time in Python : "+str(end - start))
    return()

# PASTRY module run as the main program.
if __name__ == "__main__":

    print("\n")
    print("================ PASTRY_sim v 1.0 ================ ")
    print("===== v.1.0 Seoul National University Hanlab ===== \n")

    parser = argparse.ArgumentParser()
    parser.set_defaults(method = pastry_sim)

    parser.add_argument("-n","--N", type=int, default=5, help=":# of studies combined", metavar="integer")
    parser.add_argument("-r", "--RRs", type=float, default=1.3, help=":Risk Ratio (default : 1.3)")
    parser.add_argument("-m", "--maf", type=float, default=0.3, help=":maf files directory (casemaf.out, contmaf.out) ")      
    parser.add_argument("-p", "--nrep", type=int, default=5000, help=":# of Simulation (default : 5000)")
    parser.add_argument("-s", "--shared", type=int, default=1000, help=":# of shared control", metavar="integer")
    parser.add_argument("-l", "--cont", type=int, default=1000, help=":# of control (default : 1000)")
    parser.add_argument("-e","--case", type=int, default=1000, help=":# of case (default : 1000)", metavar="integer")
    parser.add_argument("-t","--thrs", type=float, default=5E-8, help=":# of control", metavar="integer")
    options = parser.parse_args()

    pastry_sim(RRs=options.RRs, maf=options.maf, nrep=options.nrep, k=options.N, ncase=options.case, ncont=options.cont, ncontshared=options.shared, t=options.thrs)
    print("PASTRY Simulation Done. ========================= ")
    print("================================================= \n ")

File: R/PASTRY_sim.R
#!/usr/bin/env Rscript

## PASTRY and LS simulation code
## This code only consider overlapping controls between the studies(all) of the meta-analysis

## Now load or install & load all
if(!require(optparse)){
    install.packages("optparse")
    library(optparse)
}

## Option argparse
option_list = list(
    make_option(c("-n","--N"), type="integer", default=5,
              help=":# of studies combined (default : 5)", metavar="integer"), # meta-analysis studies	
    make_option(c("-r", "--RRs"), type="numeric", default=1.3, # Risk Ratio
              help=":Risk Ratio (default : 1.3)", metavar="float"),
    make_option(c("-k", "--nrep"), type="integer", default=5000, # simulation N
              help=":# of Simulation (default : 5000)", metavar="integer"),                
    make_option(c("-m", "--maf"), type="numeric", default=0.3, # MAF
              help=":MAF ", metavar="float"),         
    make_option(c("-s", "--shared"), type="integer", default=1000,
              help=":# of shared control (default : 1000)", metavar="integer"),  # shared control N
    make_option(c("-l","--cont"), type="integer", default=1000,
              help=":# of control (default : 1000)", metavar="integer"),  # control 
    make_option(c("-e","--case"), type="integer", default=1000,
              help=":# of case (default : 1000)", metavar="integer"),            # case N
    make_option(c("-t","--thrs"), type="numeric", default=5E-8,
            help=":Threshold (default : 5e-08)", metavar="float")          # threshold
); 

 
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

sim_PASTRY <- function(RRs=opt$RRs, nrep=opt$nrep, ncase=opt$case, ncont=opt$cont, ncontshared=opt$shared, nstudy=opt$N, t=opt$thrs, maf=opt$maf) {

	## Study-specific controls: ncontspe (Type: nstudy sized vector)
	## Shared controls: ncontshared (Type: nstudy sized vector
        ## Number of cases: ncase (Type: nstudy sized vector)
        ## Number of controls: ncontspe + ncontshared
	## Minor allele frequency: MAF (default=0.3)
        ## Prevalence: preval (default:0.00)

 	m=nstudy
 	maf=maf
	preval=0.000
	casenshared=0
	casensplit=casenshared/m
	contnsplit=ncontshared/m
    RRs=rep(RRs, nstudy)
    ncase=rep(ncase,nstudy)
    ncont=rep(ncont,nstudy)
    ncontspe=ncont-rep(ncontshared,nstudy)


	cat(" * Check the parameter : \n")
	cat("  - Studies # for meta-analysis : ",nstudy,"\n")
	cat("  - Risk Ratio : ",RRs,"\n")
	cat("  - MAF : ",maf,"\n")
	cat("  - Simulation # : ",nrep,"\n")
	cat("  - Case # : ",ncase,"\n")
	cat("  - Non-shared cont # : ",ncontspe,"\n")
	cat("  - Shared cont # : ",ncontshared,"\n")
	cat("  - Threshold : ",t,"\n\n")

	## Case/control frequencies
	casemaf=(RRs*maf) / ((RRs-1)*maf + 1)
	contmaf=(maf-preval*casemaf)/(1-preval)
    
	## Lin_correlation matrix
	R = diag(m)

	for (i in 1:m) {
		for (j in 1:m) {
			if (i == j) {
				next
			} else {
				first=ncontshared*sqrt(((ncase[i]*ncase[j]))/((ncontspe[i]+ncontshared)*(ncontspe[j]+ncontshared)))
				sec=casenshared*sqrt(((ncontspe[i]+ncontshared)*(ncontspe[j]+ncontshared))/((ncase[i])*(ncase[j])))
				rij=first/sqrt((ncase[i]+ncontspe[i]+ncontshared)*(ncase[j]+ncontspe[j]+ncontshared))
				R[i,j] = R[j,i] = rij
			}
		}
	}

	invR = solve(R)
    #cat(paste("\n\n Lin correlation matrix : \n ",R))

	## FE subroutine
	FE <- function(beta,stders){
		inv.var=stders**(-2)
		inv.var.beta=(inv.var %*% beta)/sum(inv.var)
		inv.var.beta.var = 1/sum(inv.var)
		new.z = inv.var.beta / sqrt(inv.var.beta.var)
		p.FE = 2*pnorm(-abs(new.z),lower.tail=sign)
		list(z=new.z,p=p.FE,beta=inv.var.beta,se=sqrt(inv.var.beta.var))
	}

	##### Overlapping Subjects #####

	pvalue.LS.FE = rep(0, nrep)
	pvalue.PASTRY.FE = rep(0, nrep)
	pvalue.split.FE = rep(0, nrep)

	# m = 2 two cohort study
	i = 1; j = 2
	
	## Now generate studies
	record=2
	for (eachrep in 1:nrep) {
		betas.nonde = rep(0, m)
		stder.nonde = rep(0, m) 
		betas.split = rep(0, m)
		stder.split = rep(0, m) 
		casemac=rep(0,m)
		contmac=rep(0,m)
		contsplitmac=rep(0,m)

		for (i in 1:m) {
			#case-specific
			casemac[i]=rbinom(1, ncase[i]*2, casemaf[i])
			#cont-specific
			contmac[i]=rbinom(1, ncontspe[i]*2, contmaf[i])
			#cont-shared
			contsplitmac[i]=rbinom(1, 2*contnsplit, contmaf[i])
		}

		contsharedmac=sum(contsplitmac)
		beta.nonde = NULL
		stderr.nonde = NULL
		beta.split = NULL
		stderr.split = NULL

		for (i in 1:m) {
			## non-splitting
			n11=casemac[i]
			n10=2*ncase[i] - casemac[i]
			n01=contmac[i]+contsharedmac
			n00=2*ncontspe[i] + 2*ncontshared - contmac[i] - contsharedmac
			betas.nonde[i] = log(n11*n00/n01/n10)
  			stder.nonde[i] = sqrt(1/n11 + 1/n00 + 1/n01 + 1/n10)
			beta.nonde = c(beta.nonde, betas.nonde[i])
			stderr.nonde = c(stderr.nonde, stder.nonde[i])

			## splitting
			n11=casemac[i]
			n10=2*ncase[i] - casemac[i]
			n01=contmac[i]+contsplitmac[i]
			n00=2*ncontspe[i] + 2*contnsplit - contmac[i] - contsplitmac[i]
			betas.split[i] = log(n11*n00/n01/n10)
			stder.split[i] = sqrt(1/n11 + 1/n00 + 1/n01 + 1/n10)
			beta.split = c(beta.split, betas.split[i])
			stderr.split = c(stderr.split, stder.split[i])
		}

	    ## Lin
		invstdmat <- solve(diag(stder.nonde))
		invsigma <- invstdmat %*% invR %*% invstdmat
		unit <- t(rep(1,nstudy))
		invvar <- unit %*% invsigma %*% t(unit)
		beta_num <- unit %*% invsigma %*% (betas.nonde)
		ls_beta <- beta_num/invvar
		ls_var <- 1/invvar
		pvalue.LS.FE[eachrep] = 2*pnorm(-abs(ls_beta/sqrt(ls_var)),lower.tail=sign)
		
        if (eachrep == 1) {cat(" * PASTRY statistics are generating... for each SNP. \n")
        start_time <- Sys.time()}
        
		######################
		i = 1; j = 2
		ntotali = ncontspe[i] + ncase[i] + ncontshared
        ntotalj = ncontspe[j] + ncase[j] + ncontshared

		p0cont=1-contmaf[i]
        p1cont=contmaf[i]
        p0case=1-casemaf[i]
        p1case=casemaf[i]

		x_0 = matrix(c(1,0,0,0),nrow=2,ncol=2)
        x_1 = matrix(c(1,1,1,1),nrow=2,ncol=2)

		info_i = matrix(c(0,0,0,0),nrow=2,ncol=2)
        info_j = matrix(c(0,0,0,0),nrow=2,ncol=2)

		up_covmat = matrix(c(0,0,0,0),nrow=2,ncol=2)
		######################

		#x <- t(betas.nonde)
        exp_alpha_i = as.numeric(exp(log(ncase[i]/(ncontspe[i]+ncontshared)) - (ls_beta*((ncase[i]*casemaf[i]+(ncontspe[i]+ncontshared)*contmaf[i])/ntotali))))
		exp_alpha_j = as.numeric(exp(log(ncase[j]/(ncontspe[j]+ncontshared)) - (ls_beta*((ncase[j]*casemaf[j]+(ncontspe[j]+ncontshared)*contmaf[j])/ntotalj))))

		x0i = exp_alpha_i ; x1i = as.numeric(exp_alpha_i*exp(ls_beta))
		x0j = exp_alpha_j ; x1j = as.numeric(exp_alpha_j*exp(ls_beta))

        ### Information Matrix for PASTRY
        info_i = 2*ncase[i]*(p0case*(x0i/(1+x0i)**2)*x_0 + p1case*(x1i/(1+x1i)**2)*x_1) + 2*(ncontspe[i]+ncontshared)*(p0cont*(x0i/(1+x0i)**2)*x_0 + p1cont*(x1i/(1+x1i)**2)*x_1)
        info_j = 2*ncase[j]*(p0case*(x0j/(1+x0j)**2)*x_0 + p1case*(x1j/(1+x1j)**2)*x_1) + 2*(ncontspe[j]+ncontshared)*(p0cont*(x0j/(1+x0j)**2)*x_0 + p1cont*(x1j/(1+x1j)**2)*x_1)

        ### Covarinace Matrix for PASTRY
        cov00=p0cont*(x0i*x0j/((1+x0i)*(1+x0j)))*x_0
        cov11=p1cont*(x1i*x1j/((1+x1i)*(1+x1j)))*x_1

        covmatrix=2*ncontshared*(cov00+cov11)
        invi = solve(info_i) ; invj = solve(info_j);
		covfinal=invi %*% covmatrix %*% invj

		#PASTRY correlation
		cor=(covfinal[2,2]/sqrt(invi[2,2]*invj[2,2]))

		#depends on study #
		PASTRY.cor=matrix(rep(cor,nstudy*nstudy),nrow=nstudy,ncol=nstudy)
		diag(PASTRY.cor)=1

 		#################################################

		## new decoupling
		invsigmaPASTRY <- invstdmat %*% solve(PASTRY.cor) %*% invstdmat
		invvarPASTRY <- unit %*% invsigmaPASTRY %*% t(unit)
	    pastry_var <- 1/invvarPASTRY
		beta_num <- unit %*% invsigmaPASTRY %*% (betas.nonde)	
        pastry_beta <- beta_num/invvarPASTRY
		
		pvalue.split.FE[eachrep] = FE(beta.split,stder.split)$p ## pvalue calculation by FE
		pvalue.PASTRY.FE[eachrep] = 2*pnorm(-abs(pastry_beta/sqrt(pastry_var)),lower.tail=sign)

		if(eachrep%%50000==0){cat(eachrep)}
		
}
        end_time <- Sys.time()
		cat(" * # Simulation : ",nrep," \n * Simulation Results : \n  LS : ",sum(pvalue.LS.FE<t)/nrep,", PASTRY",sum(pvalue.PASTRY.FE<t)/nrep,", split_FE",sum(pvalue.split.FE<t)/nrep,"\n")
        cat("\n")
        cat("PASTRY running time in R : ", end_time - start_time, ". \n")
}

cat("================ PASTRY v1.0 -- Simulation ================ \n")
sim_PASTRY()
cat("================ PASTRY Simulation END ================")
cat("\n\n")